### PR TITLE
Stash Settings: Add double key type

### PR DESCRIPTION
**Pull request by xiota**
Split from #3000.  Adds double key type to stash settings.

Example usage (`stash-example.c`):
```C
StashGroup *group;
gboolean porcelain_enabled;
gchar *potter_name;
gint stock;
gdouble price;
const gchar filename[] = "/path/data.conf";

/* setup the group */
group = stash_group_new("cup");
stash_group_add_boolean(group, &porcelain_enabled, "porcelain", TRUE);
stash_group_add_string(group, &potter_name, "potter_name", "Miss Clay");
stash_group_add_integer(group, &stock, "stock", 5);
stash_group_add_double(group, &price, "price", 1.50);

/* load the settings from a file */
if (!stash_group_load_from_file(group, filename))
	g_warning(_("Could not load keyfile %s!"), filename);

/* now use settings porcelain_enabled, potter_name, stock, and price */
/* ... */

/* save settings to file */
if (stash_group_save_to_file(group, filename, G_KEY_FILE_NONE) != 0)
	g_error(_("Could not save keyfile %s!"), filename);

/* free memory */
stash_group_free(group);
```
Generated config:
```
[cup]
porcelain=true
potter_name=Miss Clay
stock=5
price=1.5
```

### CODE REVIEW

**Comment by kugel-** [comment]
Do you have a compelling use case for this change?

**Comment by xiota** [comment]
I have a plugin that uses the double type.  Without this change, I can't use the stash system with that plugin.

File: src/utils.c [hunk 1]
```diff
@@ -795,11 +795,11 @@ gchar *utils_get_initials(const gchar *name)
  *  @param config A GKeyFile object.
  *  @param section The group name to look in for the key.
  *  @param key The key to find.
- *  @param default_value The default value which will be returned when @a section or @a key
- *         don't exist.
+ *  @param default_value The default value which will be returned when @a section
+ *         or @a key don't exist.
```
Comment (kugel-) [inline src/utils.c:799]: No unrelated changes, please.

File: src/stash.c [hunk 2]
```diff
@@ -476,6 +497,20 @@ void stash_group_add_boolean(StashGroup *group, gboolean *setting,
...
+ * @param group .
+ * @param setting Address of setting variable.
+ * @param key_name Name for key in a @c GKeyFile.
+ * @param default_value Value to use if the key doesn't exist when loading. */
+GEANY_API_SYMBOL
+void stash_group_add_double(StashGroup *group, gdouble *setting,
+		const gchar *key_name, gdouble default_value)
+{
+	gulong *default_long = (gulong*) &default_value;
+	add_pref(group, G_TYPE_DOUBLE, setting, key_name, (gpointer) *default_long);
```
Comment (kugel-) [inline src/stash.c:510]: add_pref should use a union. I'll prepare a change, then you can rebase this.

**Review by kugel- — CHANGES_REQUESTED** [review]
What does your plugin do? I can understand the desire for doubles but comments (given that keyfiles are typically not meant to be looked at by users)?

**Comment by xiota** [comment]
> given that keyfiles are typically not meant to be looked at by users

Some keyfiles are intended to be read and modified by users.  Including comments from the source makes them translatable through the gettext system.  If someone later wastes time wrapping the config with a GUI, the same comments can be repurposed as explanatory tooltips.

File: src/utils.c (hunk 1 above)
Comment (xiota) [reply src/utils.c:799]: Ok.  Will revert this when rebasing on the union code you will provide.

File: src/stash.c (hunk 2 above)
Comment (xiota) [reply src/stash.c:510]: Will rebase when it's available.  Thank you.

**Comment by xiota** [comment]
@kugel- 

I don't understand the benefit of using union over casting the pointers.  Based on what I'm reading online, they're both considered "bad".  It also looks like switching to using unions may require changing existing code, which I've been trying to avoid.

I'm probably doing this wrong...  I tried adding the following union to `stash.h`:
```C
typedef union
{
	gboolean b;
	gint n;
	gdouble d;
	gchar *s;
	gchar **v;
	const gchar *cs;
	const gchar **cv;
} StashPrefType;
```

The StashPref struct becomes:
```C
struct StashPref
{
	/* ... */
	StashPrefType *setting;
	StashPrefType default_value;
	/* ... */
};
```

Some functions need to be redefined:
```C
add_pref(StashGroup *group, GType type, StashPrefType *setting,
		const gchar *key_name, StashPrefType default_value);

static StashPref *
add_widget_pref(StashGroup *group, GType setting_type, StashPrefType *setting,
		const gchar *key_name, StashPrefType default_value,
		GType widget_type, StashWidgetID widget_id);

void stash_group_add_widget_property(StashGroup *group, StashPrefType *setting,
		const gchar *key_name, StashPrefType default_value, StashWidgetID widget_id,
		const gchar *property_name, GType type);
```

To use StashPrefType requires something like:
```C
int *setting = &se->setting->n;
double *setting = &se->setting->d;
```

Then calls to the above functions need to have casts to `StashPrefType`, like:
```C
add_pref(group, G_TYPE_BOOLEAN, (StashPrefType *) setting, key_name, (StashPrefType)default_value);
```

And any external uses of `stash_group_add_widget_property()` (`sidebar.c` and maybe plugins) also need to cast StashPrefType.

**Comment by kugel-** [comment]
See #3005, I said I prepare a change. Why did you implement the union as well?

**Comment by xiota** [comment]
> Why did you implement the union as well?

Wanted to see what the change looks like... and impatient.  I also don't understand the use of union, so comparing implementations (or getting your feedback on what I'm doing wrong) may help me understand them better.

**Comment by kugel-** [comment]
OK. So I did the change in a way that the union is only internally used and not exposed to users of the stash system.

Users call type-specific functions already, which is fine. stash_group_add_widget_property() is an exception because widget properties can be anything as well. I did not want to change that function yet. One could go and make a switch/case on the type in order to fill the union properly (again, the function signature doesn't need changing and it would be hard to do as the API is already exported).

**Comment by xiota** [comment]
@kugel- Adding the double with your changes definitely looks cleaner.

**Comment by elextr** [comment]
Just an aside

> Based on what I'm reading online, they're both considered "bad".

Indeed, but occasionally essential.  The way @kugel- did it in #3005 which encapsulated the badness, away from people using the library, is the right way.

**Comment by xiota** [comment]
@elextr Code to learn from.

**Comment by xiota** [comment]
@kugel-

Force push to rebase and change PR for the union.  Please let me know if this needs any more changes.
